oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java

```diff
@@ -0,0 +1,256 @@
+		Instant iat = (Instant) claims.get("iat");
+		Instant exp = (Instant) claims.get("exp");
+
+		// construct token
+		OAuth2TokenWithClaims withClaims = new OAuth2TokenWithClaims(bearer.getToken(), iat, exp, claims);
+		Collection<GrantedAuthority> authorities = extractAuthorities(claims);
+		AbstractAuthenticationToken result = new IntrospectedOAuth2TokenAuthenticationToken(withClaims, authorities);
+		result.setDetails(bearer.getDetails());
+		return result;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private TokenIntrospectionSuccessResponse introspect(String token) {
+		HttpHeaders headers = new HttpHeaders();
+		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON_UTF8));
+		RequestEntity<String> requestEntity =
+				new RequestEntity<>("token=" + token, headers, HttpMethod.POST, this.introspectionUri);
+		TokenIntrospectionSuccessResponse response = asSuccess(parse(adapt(request(requestEntity))));
+
+		// relying solely on the authorization server to validate this token (not checking 'exp', for example)
+		if (!response.isActive()) {
+			throw new OAuth2AuthenticationException(
+					invalidToken("Provided token [" + token + "] isn't active"));
+		}
+		return response;
+	}
+
+	private ResponseEntity<String> request(RequestEntity<String> requestEntity) {
+		try {
+			return this.restOperations.exchange(requestEntity, String.class);
+		} catch (Exception ex) {
+			throw new OAuth2AuthenticationException(
+					invalidToken(ex.getMessage()), ex);
+		}
+	}
+
+	private HTTPResponse adapt(ResponseEntity<String> responseEntity) {
+		HTTPResponse response = new HTTPResponse(responseEntity.getStatusCodeValue());
+		response.setHeader(HttpHeaders.CONTENT_TYPE, responseEntity.getHeaders().getContentType().toString());
+		response.setContent(responseEntity.getBody());
+
+		if (response.getStatusCode() != HTTPResponse.SC_OK) {
+			throw new OAuth2AuthenticationException(
+					invalidToken("Introspection endpoint responded with " + response.getStatusCode()));
+		}
+		return response;
+	}
+
+	private TokenIntrospectionResponse parse(HTTPResponse response) {
+		try {
+			return TokenIntrospectionResponse.parse(response);
+		} catch (Exception ex) {
+			throw new OAuth2AuthenticationException(
+					invalidToken(ex.getMessage()), ex);
+		}
+	}
+
+	private TokenIntrospectionSuccessResponse asSuccess(TokenIntrospectionResponse introspectionResponse) {
+		if (!introspectionResponse.indicatesSuccess()) {
+			throw new OAuth2AuthenticationException(invalidToken("Token introspection failed"));
+		}
+		return (TokenIntrospectionSuccessResponse) introspectionResponse;
+	}
+
+	private Map<String, Object> extractClaimSet(TokenIntrospectionSuccessResponse response) {
+		Map<String, Object> claims = response.toJSONObject();
+		if (response.getAudience() != null) {
```

Comment:
Why do we need to explicitly extract each claim? There may be custom (non-standard) claims in the response that we won't know and will be excluded in the claims set. I also see that `jti` and `sub` are not extracted here. Either way, I don't think we should extract claims individually. I feel all claims returned from the response should be the claim set that is composed in the `Authentication`

Reply:
We need to ensure that each claim is of the correct type, e.g. we are coercing here on write instead of requiring coercion on read. Or did I misunderstand and you are asking for something else?

Reply:
After chatting with @jgrandja offline about this, it appears that I made a poor choice in method name here. I'll update it to clarify its role.